RecoBTag/SecondaryVertex/src/CandidateBoostedDoubleSecondaryVertexComputer.cc

```diff
@@ -54,111 +95,475 @@ float CandidateBoostedDoubleSecondaryVertexComputer::discriminator(const TagInfo
-  if (tau1 != 0.) tau21 = tau2/tau1;
 
-  const std::vector<reco::CandidatePtr> & selectedTracks( ipTagInfo.selectedTracks() );
-  size_t trackSize = selectedTracks.size();
   const reco::VertexRef & vertexRef = ipTagInfo.primaryVertex();
+  GlobalPoint pv(0.,0.,0.);
+  if ( ipTagInfo.primaryVertex().isNonnull() )
+    pv = GlobalPoint(vertexRef->x(),vertexRef->y(),vertexRef->z());
+
+  const std::vector<reco::CandidatePtr> & selectedTracks = ipTagInfo.selectedTracks();
+  const std::vector<reco::btag::TrackIPData> & ipData = ipTagInfo.impactParameterData();
+  size_t trackSize = selectedTracks.size();
+
+
   reco::TrackKinematics allKinematics;
+  std::vector<float> IP3Ds, IP3Ds_1, IP3Ds_2;
+  int contTrk=0;
 
+  // loop over tracks associated to the jet
   for (size_t itt=0; itt < trackSize; ++itt)
   {
-    const reco::Track & ptrack = *(reco::btag::toTrack(selectedTracks[itt]));
     const reco::CandidatePtr ptrackRef = selectedTracks[itt];
+    const reco::Track * ptrackPtr = reco::btag::toTrack(ptrackRef);
+    const reco::Track & ptrack = *ptrackPtr;
 
     float track_PVweight = 0.;
     setTracksPV(ptrackRef, vertexRef, track_PVweight);
-    if (track_PVweight>0.) { allKinematics.add(ptrack, track_PVweight); }
+    if (track_PVweight>0.5) allKinematics.add(ptrack, track_PVweight);
+
+    const reco::btag::TrackIPData &data = ipData[itt];
+    bool isSelected = false;
+    if (trackSelector(ptrack, data, *jet, pv)) isSelected = true;
+
+    // check if the track is from V0
+    bool isfromV0 = false, isfromV0Tight = false;
+    const reco::Track * trackPairV0Test[2];
+
+    trackPairV0Test[0] = ptrackPtr;
+
+    for (size_t jtt=0; jtt < trackSize; ++jtt)
+    {
+      if (itt == jtt) continue;
+
+      const reco::btag::TrackIPData & pairTrackData = ipData[jtt];
+      const reco::CandidatePtr pairTrackRef = selectedTracks[jtt];
+      const reco::Track * pairTrackPtr = reco::btag::toTrack(pairTrackRef);
+      const reco::Track & pairTrack = *pairTrackPtr;
+
+      trackPairV0Test[1] = pairTrackPtr;
+
+      if (!trackPairV0Filter(trackPairV0Test, 2))
+      {
+        isfromV0 = true;
+
+        if ( trackSelector(pairTrack, pairTrackData, *jet, pv) )
+          isfromV0Tight = true;
+      }
+
+      if (isfromV0 && isfromV0Tight)
+        break;
+    }
+
+    if( isSelected && !isfromV0Tight ) jetNTracks += 1.;
+
+    reco::TransientTrack transientTrack = trackBuilder->build(ptrack);
+    GlobalVector direction(jet->px(), jet->py(), jet->pz());
+
+    if (currentAxes.size() > 1)
+    {
+      if (reco::deltaR2(ptrack,currentAxes[1]) < reco::deltaR2(ptrack,currentAxes[0]))
+        direction = GlobalVector(currentAxes[1].px(), currentAxes[1].py(), currentAxes[1].pz());
+      else
+        direction = GlobalVector(currentAxes[0].px(), currentAxes[0].py(), currentAxes[0].pz());
+    }
+    else if (currentAxes.size() > 0)
+      direction = GlobalVector(currentAxes[0].px(), currentAxes[0].py(), currentAxes[0].pz());
```

Comment:
Much simpler and to avoid code duplication:
```
int index = 0;
if (currentAxes.size() > 1 && reco::deltaR2(ptrack,currentAxes[1]) < reco::deltaR2(ptrack,currentAxes[0]))
    index = 1;
direction = GlobalVector(currentAxes[index].px(), currentAxes[index].py(), currentAxes[index].pz());
```